lib/remote/apilistener-filesync.cpp

```diff
@@ -312,7 +312,15 @@ Value ApiListener::ConfigUpdateHandler(const MessageOrigin::Ptr& origin, const D
 		return Empty;
 	}
 
-	std::thread([origin, params]() { HandleConfigUpdate(origin, params); }).detach();
+	std::thread([origin, params]() {
+		try {
+			HandleConfigUpdate(origin, params);
+		} catch (const std::exception& e) {
+			Log(LogCritical, "ApiListener")
+				<< "Exception during HandleConfigUpdate: " << e.what();
+			Application::Exit(EXIT_FAILURE);
```

Comment:
What's this for?

Reply:
Why it exits? Because any error might have occurred and I don't see that there's a way to safely continue in this case. This change is not supposed to fix a crash but to get better log message to help diagnosing what caused it.

Reply:
> I don't see that there's a way to safely continue in this case.

You just... continue?

Reply:
Already discussed that with @N-o-X, this will result in a somewhat broken node that no longer receives config updates.

Reply:
Why exactly?

Reply:
I don't think just continuing and logging a critical isn't enough here. Most users will never see that critical, if it appears. They will just have a broken cluster and don't know about it until some checks are overdue because they didn't get synced.

Reply:
Thinking about this again, it could be the case that it's just a transient error that goes away by itself. I don't know all the details of the config update to be confident that it's safe to just continue for any exception, however this is probably what happens anyways when you restart the node.

Reply:
In my opinion logging isn't enough here. But instead of killing the process, we could also use something similar to `SetLastFailedZonesStageValidation()` and tell the user that something went wrong via the Icinga check.

Reply:
I agree.

Reply:
On what? Together with @lippserd we agreed on killing the process, because else many users might not realise that something is wrong.

Reply:
On https://github.com/Icinga/icinga2/pull/8318#discussion_r502243357

Reply:
And I strongly disagree with killing the process. A critical log message **and** `SetLastFailedZonesStageValidation()` is enough.

Reply:
This is more of a quick fix to get at least some useful information from the exception. A nicer fix will follow at a later time.

Reply:
As far as I remember we **always** catch and log our exceptions and **continue**. Why not here?

Reply:
This will leave you with a node running some old config and we have no idea if it is a persistent or transient error. And even if it is a transient error, you don't know when the next config sync would happen that could fix the problem. So we agreed that it is better to fail in a way that an admin would notice.

Reply:
Look at v2.11.0. Our config update handler's exceptions...

https://github.com/Icinga/icinga2/blob/v2.11.0/lib/remote/apilistener-filesync.cpp#L244-L510

... are caught by JsonRpcConnection:

https://github.com/Icinga/icinga2/blob/v2.11.0/lib/remote/jsonrpcconnection.cpp#L80-L92

We have broken this and should repair it.

Reply:
Stopping Icinga here was originally my idea. At second glance, I would also say that we should better catch the exception, then mark it in the log and with `SetLastFailedZonesStageValidation()` as with the other errors in the config update.